pkg/cmd/infra/router/template.go

```diff
@@ -287,6 +293,102 @@ func parseCaptureHeaders(in string) ([]templateplugin.CaptureHTTPHeader, error)
+	var captureHeaders []templateplugin.HTTPHeader
+	var capture templateplugin.HTTPHeader
+	var err error
+	if len(in) == 0 {
+		return captureHeaders, fmt.Errorf("encoded header string not present")
+	}
+	if len(in) > 0 {
+		for _, header := range strings.Split(in, ",") {
+			parts := strings.Split(header, ":")
+			num := len(parts)
+			switch num {
+			default:
+				return captureHeaders, fmt.Errorf("invalid HTTP header input specification: %v", header)
+			case 3:
+				{
+					headerName, err := url.QueryUnescape(parts[0])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[0])
+					}
+					err = checkValidHeaderName(headerName)
+					if err != nil {
+						return captureHeaders, err
+					}
+					headerValue, err := url.QueryUnescape(parts[1])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[1])
+					}
+					sanitizedHeaderValue := templateplugin.SanitizeHeaderValue(headerValue)
+					action, err := url.QueryUnescape(parts[2])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[2])
+					}
+					err = checkValidAction(action)
+					if err != nil {
+						return captureHeaders, err
+					}
+					capture = templateplugin.HTTPHeader{
+						Name:   headerName,
+						Value:  sanitizedHeaderValue,
+						Action: routev1.RouteHTTPHeaderActionType(action),
+					}
+					captureHeaders = append(captureHeaders, capture)
+				}
+			case 2:
+				{
+					headerName, err := url.QueryUnescape(parts[0])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[0])
+					}
+					err = checkValidHeaderName(headerName)
+					if err != nil {
+						return captureHeaders, err
+					}
+					action, err := url.QueryUnescape(parts[1])
+					if err != nil {
+						return captureHeaders, fmt.Errorf("failed to decode percent encoding: %v", parts[1])
+					}
+					err = checkValidAction(action)
+					if err != nil {
+						return captureHeaders, err
+					}
+					capture = templateplugin.HTTPHeader{
+						Name:   headerName,
+						Action: routev1.RouteHTTPHeaderActionType(action),
+					}
+					captureHeaders = append(captureHeaders, capture)
+				}
+			}
+		}
+	}
+
+	return captureHeaders, err
+}
+
+func checkValidAction(action string) error {
+	if action != string(routev1.Set) && action != string(routev1.Delete) {
+		return fmt.Errorf("invalid action: %s", action)
+	} else {
+		return nil
+	}
+}
+
+func checkValidHeaderName(headerName string) error {
+	// RFC 2616, section 4.2, states that the header name
+	// must be a valid token.
+	// Any changes made to regex of header name in route type in openshift/api and `validation.go` file in library-go shall be reflected here and
+	// vice-versa
+	permittedHeaderNameRE := regexp.MustCompile("^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$")
```

Comment:
Don't recompile the regexp every time the function is called.  Don't use the word "shall".  End your sentences with a period.  Remember to add godoc. 
```suggestion
// permittedHeaderNameRE is a compiled regexp to match an HTTP header name
// as specified in RFC 2616, section 4.2.
// Any changes made to regex of header name in route type in openshift/api and `validation.go` file in library-go must be reflected here and
// vice versa.
var permittedHeaderNameRE = regexp.MustCompile("^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$")

// checkValidHeaderName verifies that the given HTTP header name is valid.
func checkValidHeaderName(headerName string) error {
```